ccl/streamingccl/streamproducer: TestStreamPartition failed

ccl/streamingccl/streamproducer.TestStreamPartition [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4186481&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4186481&tab=artifacts#/) on master @ [e1068d77afbd39b162978281c9da7cbea49c1c3a](https://github.com/cockroachdb/cockroach/commits/e1068d77afbd39b162978281c9da7cbea49c1c3a):


```
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/log.go:82 +0x88d
  github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).logSplit()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/log.go:106 +0x3cc
  github.com/cockroachdb/cockroach/pkg/kv/kvserver.splitTxnAttempt()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_command.go:210 +0x3d7
  github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).adminSplitWithDescriptor.func2()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_command.go:404 +0x108
  github.com/cockroachdb/cockroach/pkg/kv.runTxn.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/db.go:872 +0x51
  github.com/cockroachdb/cockroach/pkg/kv.(*Txn).exec()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/txn.go:972 +0xc1
  github.com/cockroachdb/cockroach/pkg/kv.runTxn()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/db.go:871 +0x84
  github.com/cockroachdb/cockroach/pkg/kv.(*DB).Txn()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/db.go:853 +0xbd
  github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).adminSplitWithDescriptor()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_command.go:403 +0xd24
  github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*splitQueue).processAttempt()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/split_queue.go:184 +0x1e6
  github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*splitQueue).process()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/split_queue.go:164 +0x96
  github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*baseQueue).processReplica.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/queue.go:966 +0x282
  github.com/cockroachdb/cockroach/pkg/util/contextutil.RunWithTimeout()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/contextutil/context.go:91 +0xed
  github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*baseQueue).processReplica()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/queue.go:925 +0x564
  github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*baseQueue).processLoop.func2.1()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/queue.go:844 +0x104
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:494 +0x235

Goroutine 533 (running) created at:
  github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).processCommandsAsync()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:597 +0x404
  github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).serveImpl()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:324 +0xc93
  github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*Server).serveConn()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:171 +0x4b2
  github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*Server).ServeConn()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/server.go:712 +0xf0f
  github.com/cockroachdb/cockroach/pkg/server.(*SQLServer).startServeSQL.func1.1()
      /go/src/github.com/cockroachdb/cockroach/pkg/server/server.go:2545 +0x109
  github.com/cockroachdb/cockroach/pkg/util/netutil.(*Server).ServeWith.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/netutil/net.go:162 +0x244
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:494 +0x235
==================
    testing.go:1152: race detected during execution of test
    --- FAIL: TestStreamPartition/stream-batches-events (1.28s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
Parameters in this failure:

- GOFLAGS=-race -parallel=4
</p>
</details>
/cc @cockroachdb/cdc
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStreamPartition.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

ccl/streamingccl/streamproducer.TestStreamPartition [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4189926&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4189926&tab=artifacts#/) on master @ [8cd28089c6c7333615ba3201e841839001d2f0e1](https://github.com/cockroachdb/cockroach/commits/8cd28089c6c7333615ba3201e841839001d2f0e1):


```
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/rowexec/project_set.go:158 +0x535
  github.com/cockroachdb/cockroach/pkg/sql/rowexec.(*projectSetProcessor).Next()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/rowexec/project_set.go:241 +0x8b
  github.com/cockroachdb/cockroach/pkg/sql/colexec.(*Columnarizer).Next()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/columnarizer.go:229 +0x4ce
  github.com/cockroachdb/cockroach/pkg/sql/colexec.(*invariantsChecker).Next()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/invariants_checker.go:94 +0x6f
  github.com/cockroachdb/cockroach/pkg/sql/colflow.(*batchInfoCollector).Next()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/stats.go:99 +0x65
  github.com/cockroachdb/cockroach/pkg/sql/colflow.(*vectorizedStatsCollectorImpl).Next()
      <autogenerated>:1 +0x37
  github.com/cockroachdb/cockroach/pkg/sql/colflow.(*BatchFlowCoordinator).nextAdapter()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/flow_coordinator.go:240 +0x5e
  github.com/cockroachdb/cockroach/pkg/sql/colflow.(*BatchFlowCoordinator).nextAdapter-fm()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/flow_coordinator.go:239 +0x29
  github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:91 +0x7b
  github.com/cockroachdb/cockroach/pkg/sql/colflow.(*BatchFlowCoordinator).next()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/flow_coordinator.go:244 +0x3db
  github.com/cockroachdb/cockroach/pkg/sql/colflow.(*BatchFlowCoordinator).Run()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/flow_coordinator.go:283 +0x3a6
  github.com/cockroachdb/cockroach/pkg/sql/colflow.(*vectorizedFlow).Run()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/vectorized_flow.go:259 +0x328
  github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:563 +0xf94
  github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRun()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1401 +0x249
  github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithDistSQLEngine()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1408 +0xceb
  github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1085 +0x12c4
  github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:722 +0x342a
  github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:135 +0x93c
  github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execPortal()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:222 +0x1dc
  github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd.func2()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1820 +0xa1c
  github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1822 +0x6f8
  github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1676 +0x5fe
  github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:681 +0x14b
  github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).processCommandsAsync.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:669 +0x405
==================
    testing.go:1152: race detected during execution of test
    --- FAIL: TestStreamPartition/stream-table-with-cursor (1.03s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
Parameters in this failure:

- GOFLAGS=-race -parallel=4
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStreamPartition.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

ccl/streamingccl/streamproducer.TestStreamPartition [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4189925&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4189925&tab=artifacts#/) on master @ [8cd28089c6c7333615ba3201e841839001d2f0e1](https://github.com/cockroachdb/cockroach/commits/8cd28089c6c7333615ba3201e841839001d2f0e1):


```
=== RUN   TestStreamPartition
    test_log_scope.go:79: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestStreamPartition2136112011
    test_log_scope.go:80: use -show-logs to present logs inline
=== CONT  TestStreamPartition
    replication_stream_test.go:481: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestStreamPartition2136112011
--- FAIL: TestStreamPartition (4.78s)
=== RUN   TestStreamPartition/stream-table-with-cursor
    replication_stream_test.go:438: 
        	Error Trace:	replication_stream_test.go:438
        	Error:      	Not equal: 
        	            	expected: []byte{0x0, 0x8e, 0x5d, 0xb5, 0xa, 0x26, 0xc, 0xd0, 0xbf, 0xd1, 0x80, 0xd0, 0xb8, 0xd0, 0xb2, 0xd0, 0xb5, 0xd1, 0x82, 0x16, 0x6, 0xd0, 0xbc, 0xd0, 0xb8, 0xd1, 0x80}
        	            	actual  : []byte{0x75, 0xe6, 0x2c, 0xef, 0xa, 0x26, 0xc, 0xd0, 0xbf, 0xd1, 0x80, 0xd0, 0xb8, 0xd0, 0xb2, 0xd0, 0xb5, 0xd1, 0x82, 0x16, 0x5, 0x77, 0x6f, 0x72, 0x6c, 0x64}
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1,4 +1,4 @@
        	            	-([]uint8) (len=27) {
        	            	- 00000000  00 8e 5d b5 0a 26 0c d0  bf d1 80 d0 b8 d0 b2 d0  |..]..&..........|
        	            	- 00000010  b5 d1 82 16 06 d0 bc d0  b8 d1 80                 |...........|
        	            	+([]uint8) (len=26) {
        	            	+ 00000000  75 e6 2c ef 0a 26 0c d0  bf d1 80 d0 b8 d0 b2 d0  |u.,..&..........|
        	            	+ 00000010  b5 d1 82 16 05 77 6f 72  6c 64                    |.....world|
        	            	 }
        	Test:       	TestStreamPartition/stream-table-with-cursor
    --- FAIL: TestStreamPartition/stream-table-with-cursor (1.15s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
Parameters in this failure:

- TAGS=deadlock

- GOFLAGS=-parallel=4
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStreamPartition.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

ccl/streamingccl/streamproducer.TestStreamPartition [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4224697&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4224697&tab=artifacts#/) on master @ [83e2df701688745fe7e7d8a0d0e5e7a4ba8633c8](https://github.com/cockroachdb/cockroach/commits/83e2df701688745fe7e7d8a0d0e5e7a4ba8633c8):


```
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1833 +0x72a
  github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1676 +0x5fe
  github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:681 +0x14b
  github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).processCommandsAsync.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:680 +0x405

Goroutine 623 (running) created at:
  github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).initConnEx()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/internal.go:204 +0xaa8
  github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).execInternal()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/internal.go:732 +0xcd6
  github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).ExecEx()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/internal.go:529 +0x24d
  github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).insertRangeLogEvent()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/log.go:82 +0x88d
  github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).logSplit()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/log.go:106 +0x3cc
  github.com/cockroachdb/cockroach/pkg/kv/kvserver.splitTxnAttempt()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_command.go:210 +0x3d7
  github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).adminSplitWithDescriptor.func2()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_command.go:404 +0x108
  github.com/cockroachdb/cockroach/pkg/kv.runTxn.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/db.go:874 +0x51
  github.com/cockroachdb/cockroach/pkg/kv.(*Txn).exec()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/txn.go:972 +0xc1
  github.com/cockroachdb/cockroach/pkg/kv.runTxn()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/db.go:873 +0x84
  github.com/cockroachdb/cockroach/pkg/kv.(*DB).Txn()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/db.go:855 +0xbd
  github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).adminSplitWithDescriptor()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_command.go:403 +0xd24
  github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*splitQueue).processAttempt()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/split_queue.go:184 +0x1e6
  github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*splitQueue).process()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/split_queue.go:164 +0x96
  github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*baseQueue).processReplica.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/queue.go:966 +0x282
  github.com/cockroachdb/cockroach/pkg/util/contextutil.RunWithTimeout()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/contextutil/context.go:91 +0xed
  github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*baseQueue).processReplica()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/queue.go:925 +0x564
  github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*baseQueue).processLoop.func2.1()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/queue.go:844 +0x104
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:494 +0x235
==================
    testing.go:1152: race detected during execution of test
    --- FAIL: TestStreamPartition/stream-batches-events (1.54s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
Parameters in this failure:

- GOFLAGS=-race -parallel=4
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStreamPartition.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

ccl/streamingccl/streamproducer.TestStreamPartition [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4251106&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4251106&tab=artifacts#/) on master @ [a5158c4c93e9dca858b38ef7e94321f0a0a2b5c5](https://github.com/cockroachdb/cockroach/commits/a5158c4c93e9dca858b38ef7e94321f0a0a2b5c5):


```
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/rowexec/project_set.go:158 +0x535
  github.com/cockroachdb/cockroach/pkg/sql/rowexec.(*projectSetProcessor).Next()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/rowexec/project_set.go:241 +0x8b
  github.com/cockroachdb/cockroach/pkg/sql/colexec.(*Columnarizer).Next()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/columnarizer.go:229 +0x4ce
  github.com/cockroachdb/cockroach/pkg/sql/colexec.(*invariantsChecker).Next()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/invariants_checker.go:94 +0x6f
  github.com/cockroachdb/cockroach/pkg/sql/colflow.(*batchInfoCollector).Next()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/stats.go:99 +0x65
  github.com/cockroachdb/cockroach/pkg/sql/colflow.(*vectorizedStatsCollectorImpl).Next()
      <autogenerated>:1 +0x37
  github.com/cockroachdb/cockroach/pkg/sql/colflow.(*BatchFlowCoordinator).nextAdapter()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/flow_coordinator.go:240 +0x5e
  github.com/cockroachdb/cockroach/pkg/sql/colflow.(*BatchFlowCoordinator).nextAdapter-fm()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/flow_coordinator.go:239 +0x29
  github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:91 +0x7b
  github.com/cockroachdb/cockroach/pkg/sql/colflow.(*BatchFlowCoordinator).next()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/flow_coordinator.go:244 +0x37c
  github.com/cockroachdb/cockroach/pkg/sql/colflow.(*BatchFlowCoordinator).Run()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/flow_coordinator.go:287 +0x347
  github.com/cockroachdb/cockroach/pkg/sql/colflow.(*vectorizedFlow).Run()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/vectorized_flow.go:259 +0x328
  github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:573 +0x10bc
  github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRun()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1409 +0x249
  github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithDistSQLEngine()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1400 +0xceb
  github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1077 +0x12c4
  github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:715 +0x34ca
  github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:137 +0x948
  github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execPortal()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:225 +0x1e5
  github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd.func2()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1820 +0xab7
  github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1822 +0x7bd
  github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1675 +0x468
  github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:680 +0x14b
  github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).processCommandsAsync.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:680 +0x405
==================
    testing.go:1152: race detected during execution of test
    --- FAIL: TestStreamPartition/stream-table-with-cursor (1.03s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
Parameters in this failure:

- GOFLAGS=-race -parallel=4
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStreamPartition.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

ccl/streamingccl/streamproducer.TestStreamPartition [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4264032&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4264032&tab=artifacts#/) on master @ [95c82a7dbc6d6fecb63edbd666c9986d1ad3885b](https://github.com/cockroachdb/cockroach/commits/95c82a7dbc6d6fecb63edbd666c9986d1ad3885b):


```
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:91 +0x7b
  github.com/cockroachdb/cockroach/pkg/sql/colflow.(*BatchFlowCoordinator).next()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/flow_coordinator.go:244 +0x37c
  github.com/cockroachdb/cockroach/pkg/sql/colflow.(*BatchFlowCoordinator).Run()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/flow_coordinator.go:287 +0x347
  github.com/cockroachdb/cockroach/pkg/sql/colflow.(*vectorizedFlow).Run()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/vectorized_flow.go:259 +0x328
  github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:573 +0x10bc
  github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRun()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1409 +0x249
  github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithDistSQLEngine()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1400 +0xceb
  github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1077 +0x12c4
  github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:715 +0x34ca
  github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:137 +0x948
  github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execPortal()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:225 +0x1e5
  github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd.func2()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1820 +0xab7
  github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1822 +0x7bd
  github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1675 +0x468
  github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:680 +0x14b
  github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).processCommandsAsync.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:680 +0x405

Goroutine 279 (running) created at:
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:483 +0x615
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTask()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:354 +0x188
  github.com/cockroachdb/cockroach/pkg/internal/client/requestbatcher.(*RequestBatcher).sendBatch()
      /go/src/github.com/cockroachdb/cockroach/pkg/internal/client/requestbatcher/batcher.go:273 +0x53
  github.com/cockroachdb/cockroach/pkg/internal/client/requestbatcher.(*RequestBatcher).run.func2()
      /go/src/github.com/cockroachdb/cockroach/pkg/internal/client/requestbatcher/batcher.go:425 +0x5bc
  github.com/cockroachdb/cockroach/pkg/internal/client/requestbatcher.(*RequestBatcher).run()
      /go/src/github.com/cockroachdb/cockroach/pkg/internal/client/requestbatcher/batcher.go:475 +0x539
  github.com/cockroachdb/cockroach/pkg/internal/client/requestbatcher.(*RequestBatcher).run-fm()
      /go/src/github.com/cockroachdb/cockroach/pkg/internal/client/requestbatcher/batcher.go:396 +0x4d
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:494 +0x235
==================
    testing.go:1152: race detected during execution of test
    --- FAIL: TestStreamPartition/stream-table-with-cursor (1.02s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
Parameters in this failure:

- GOFLAGS=-race -parallel=4
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStreamPartition.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

ccl/streamingccl/streamproducer.TestStreamPartition [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4288610&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4288610&tab=artifacts#/) on master @ [f20c1a8d824f0d00f18c0d58ed8b4bc4955e0969](https://github.com/cockroachdb/cockroach/commits/f20c1a8d824f0d00f18c0d58ed8b4bc4955e0969):


```
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1090 +0x12c4
  github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:720 +0x33ee
  github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:139 +0x948
  github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execPortal()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:227 +0x1e5
  github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd.func2()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1847 +0xab7
  github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1849 +0x7bd
  github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1702 +0x468
  github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:700 +0x14b
  github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).processCommandsAsync.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:680 +0x405

Goroutine 63 (running) created at:
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:483 +0x615
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTask()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:354 +0x23a
  github.com/cockroachdb/cockroach/pkg/sql/sqlliveness/slinstance.(*Instance).Start()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/sqlliveness/slinstance/slinstance.go:301 +0x26
  github.com/cockroachdb/cockroach/pkg/sql/sqlliveness/slprovider.(*provider).Start()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/sqlliveness/slprovider/slprovider.go:50 +0x75
  github.com/cockroachdb/cockroach/pkg/server.(*SQLServer).startSQLLivenessAndInstanceProviders()
      /go/src/github.com/cockroachdb/cockroach/pkg/server/server_sql.go:1019 +0xc7
  github.com/cockroachdb/cockroach/pkg/server.(*SQLServer).preStart()
      /go/src/github.com/cockroachdb/cockroach/pkg/server/server_sql.go:1060 +0x92
  github.com/cockroachdb/cockroach/pkg/server.(*Server).PreStart()
      /go/src/github.com/cockroachdb/cockroach/pkg/server/server.go:1433 +0x44c4
  github.com/cockroachdb/cockroach/pkg/server.(*Server).Start()
      /go/src/github.com/cockroachdb/cockroach/pkg/server/server.go:850 +0x44
  github.com/cockroachdb/cockroach/pkg/server.(*TestServer).Start()
      /go/src/github.com/cockroachdb/cockroach/pkg/server/testserver.go:500 +0x64
  github.com/cockroachdb/cockroach/pkg/testutils/serverutils.StartServer()
      /go/src/github.com/cockroachdb/cockroach/pkg/testutils/serverutils/test_server_shim.go:241 +0x17b
  github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingtest.NewReplicationHelper()
      /go/src/github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingtest/replication_helpers.go:190 +0x84
  github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamproducer_test.TestStreamPartition()
      /go/src/github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamproducer/replication_stream_test.go:356 +0x12e
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1259 +0x22f
  testing.(*T).Run·dwrap·21()
      /usr/local/go/src/testing/testing.go:1306 +0x47
==================
    testing.go:1152: race detected during execution of test
    --- FAIL: TestStreamPartition/stream-batches-events (1.32s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
Parameters in this failure:

- GOFLAGS=-race -parallel=4
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStreamPartition.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/ccl/streamingccl/streamproducer/BUILD.bazel
+++ b/pkg/ccl/streamingccl/streamproducer/BUILD.bazel
@@ -45,6 +45,7 @@ go_library(
         "//pkg/util/protoutil",
         "//pkg/util/span",
         "//pkg/util/timeutil",
+        "//pkg/util/tracing",
         "//pkg/util/uuid",
         "@com_github_cockroachdb_errors//:errors",
     ],

--- a/pkg/ccl/streamingccl/streamproducer/event_stream.go
+++ b/pkg/ccl/streamingccl/streamproducer/event_stream.go
@@ -28,6 +28,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/util/protoutil"
 	"github.com/cockroachdb/cockroach/pkg/util/span"
 	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
+	"github.com/cockroachdb/cockroach/pkg/util/tracing"
 	"github.com/cockroachdb/errors"
 )
 
@@ -46,6 +47,7 @@ type eventStream struct {
 	eventsCh    chan roachpb.RangeFeedEvent // Channel receiving rangefeed events.
 	errCh       chan error                  // Signaled when error occurs in rangefeed.
 	streamCh    chan tree.Datums            // Channel signaled to forward datums to consumer.
+	sp          *tracing.Span               // Span representing the lifetime of the eventStream.
 }
 
 var _ tree.ValueGenerator = (*eventStream)(nil)
@@ -171,7 +173,9 @@ func (s *eventStream) startStreamProcessor(ctx context.Context, frontier *span.F
 	// Context group responsible for coordinating rangefeed event production with
 	// ValueGenerator implementation that consumes rangefeed events and forwards them to the
 	// destination cluster consumer.
-	s.streamGroup = ctxgroup.WithContext(ctx)
+	streamCtx, sp := tracing.ChildSpan(ctx, "event stream")
+	s.sp = sp
+	s.streamGroup = ctxgroup.WithContext(streamCtx)
 	s.streamGroup.GoCtx(withErrCapture(func(ctx context.Context) error {
 		return s.streamLoop(ctx, frontier)
 	}))
@@ -206,6 +210,8 @@ func (s *eventStream) Close(ctx context.Context) {
 		// Note: error in close is normal; we expect to be terminated with context canceled.
 		log.Errorf(ctx, "partition stream %d terminated with error %v", s.streamID, err)
 	}
+
+	s.sp.Finish()
 }
 
 func (s *eventStream) onEvent(ctx context.Context, value *roachpb.RangeFeedValue) {

--- a/pkg/sql/execinfra/processorsbase.go
+++ b/pkg/sql/execinfra/processorsbase.go
@@ -892,7 +892,16 @@ func (pb *ProcessorBaseNoHelper) startImpl(
 //     // Perform processor specific close work.
 //   }
 func (pb *ProcessorBase) InternalClose() bool {
-	closing := pb.ProcessorBaseNoHelper.InternalClose()
+	return pb.InternalCloseEx(nil /* onClose */)
+}
+
+// InternalCloseEx is like InternalClose, but also takes a closure to run in
+// case the processor was not already closed. The closure is run before the
+// processor's span is finished, so the closure can finalize work that relies on
+// that span (e.g. async work previously started by the processor that has
+// captured the processor's span).
+func (pb *ProcessorBase) InternalCloseEx(onClose func()) bool {
+	closing := pb.ProcessorBaseNoHelper.InternalCloseEx(onClose)
 	if closing {
 		// This prevents Next() from returning more rows.
 		pb.OutputHelper.consumerClosed()
@@ -902,24 +911,33 @@ func (pb *ProcessorBase) InternalClose() bool {
 
 // InternalClose is the meat of ProcessorBase.InternalClose.
 func (pb *ProcessorBaseNoHelper) InternalClose() bool {
-	closing := !pb.Closed
+	return pb.InternalCloseEx(nil /* onClose */)
+}
+
+// InternalCloseEx is the meat of ProcessorBase.InternalCloseEx.
+func (pb *ProcessorBaseNoHelper) InternalCloseEx(onClose func()) bool {
 	// Protection around double closing is useful for allowing ConsumerClosed() to
 	// be called on processors that have already closed themselves by moving to
 	// StateTrailingMeta.
-	if closing {
-		for _, input := range pb.inputsToDrain[pb.curInputToDrain:] {
-			input.ConsumerClosed()
-		}
-
-		pb.Closed = true
-		pb.span.Finish()
-		pb.span = nil
-		// Reset the context so that any incidental uses after this point do not
-		// access the finished span.
-		pb.Ctx = pb.origCtx
-		pb.EvalCtx.Context = pb.origCtx
+	if pb.Closed {
+		return false
 	}
-	return closing
+	for _, input := range pb.inputsToDrain[pb.curInputToDrain:] {
+		input.ConsumerClosed()
+	}
+
+	if onClose != nil {
+		onClose()
+	}
+
+	pb.Closed = true
+	pb.span.Finish()
+	pb.span = nil
+	// Reset the context so that any incidental uses after this point do not
+	// access the finished span.
+	pb.Ctx = pb.origCtx
+	pb.EvalCtx.Context = pb.origCtx
+	return true
 }
 
 // ConsumerDone is part of the RowSource interface.

--- a/pkg/sql/rowexec/project_set.go
+++ b/pkg/sql/rowexec/project_set.go
@@ -285,13 +285,13 @@ func (ps *projectSetProcessor) toEncDatum(d tree.Datum, colIdx int) rowenc.EncDa
 }
 
 func (ps *projectSetProcessor) close() {
-	if ps.InternalClose() {
+	ps.InternalCloseEx(func() {
 		for _, gen := range ps.gens {
 			if gen != nil {
 				gen.Close(ps.Ctx)
 			}
 		}
-	}
+	})
 }
 
 // ConsumerClosed is part of the RowSource interface.

